lib/new_relic/telemetry_sdk/clients/client.rb

```diff
@@ -0,0 +1,84 @@
+# encoding: utf-8
+# frozen_string_literal: true
+# This file is distributed under New Relic's license terms.
+# See https://github.com/newrelic/newrelic-telemetry-sdk-ruby/blob/main/LICENSE for complete details.
+
+require 'net/http'
+require 'json'
+require 'zlib'
+require 'securerandom'
+
+require 'new_relic/telemetry_sdk/buffer'
+
+module NewRelic
+  module TelemetrySdk
+    class Client
+      def initialize host:,
+                     path:,
+                     headers: {},
+                     use_gzip: true,
+                     payload_type:
+        @connection = set_up_connection host
+        @path = path
+        @headers = headers
+        @gzip_request = use_gzip
+        @payload_type = payload_type
+        add_content_encoding_header @headers if @gzip_request
+      end
+
+      def send_request body
+        body = serialize body
+        body = gzip_data body if @gzip_request
+        @connection.post @path, body, @headers
+      end
+
+      def report item
+        report_batch [item]
+      end
+
+      def report_batch data, common_attributes=nil
+        # We need to generate a version 4 uuid that will
+        # be used for each unique batch, including on retries.
+        # If a batch is split due to a 413 response,
+        # each smaller batch should have its own.
+        @headers[:'x-request-id'] = SecureRandom.uuid
+
+        post_body = format_payload data, common_attributes
+        response = send_request post_body
+
+        return if response.is_a? Net::HTTPSuccess
+        # Otherwise, take appropriate action based on response code
+      end
+
+      def format_payload data, common_attributes
+        post_body = { @payload_type => data }
+
+        if common_attributes
+          post_body[:common] = {}
+          post_body[:common][:attributes] = common_attributes
+        end
+
+        [post_body]
+      end
+
+      def add_content_encoding_header headers
+        headers.merge!(:'content-encoding' => 'gzip')
+      end
+
+      def set_up_connection host
+        uri = URI(host)
+        conn = Net::HTTP.new uri.host, uri.port
+        conn.use_ssl = true
+        conn
+      end
+
+      def serialize data
+        JSON.generate data
```

Comment:
`JSON.dump` may be faster than `JSON.generate`

Reply:
Thanks for bringing this up. I did some benchmarking, and `generate` was faster - though not significantly faster.

![image](https://user-images.githubusercontent.com/10822861/94846615-475b9100-03d6-11eb-8231-7d9d58a04ccf.png)

Reply:
Nice!  So they've improved #generate over the years significantly